### PR TITLE
Implement intrinsic `allocated`

### LINKED ISSUE #1353: Implement intrinsic `allocated`
```fortran
program test_hybrd

    implicit none
    real,dimension(:),allocatable :: fvec

    if (allocated(fvec)) deallocate(fvec)

    contains
end program test_hybrd
```

> LFortran

```console
$ lfortran a.f90 --show-llvm
code generation error: Intrinsic 'allocated' not implemented yet and compile time value is not available.


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed)
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1353.
Once this gets merged, I will open up PR to implement `deallocate`. We already have implementation for that (ImplicitDeallocate).

### CODE REVIEW

File: integration_tests/allocated_01.f90 [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+    implicit none
+    real(4), allocatable :: x(:)
+
+
+    if (allocated(x)) then
+        print *, "x is allocated"
+    else 
+        print *, "x is not allocated"
+    end if
```
Comment (certik) [inline integration_tests/allocated_01.f90:11]: Add a test like `if (allocated(x)) error stop`.

File: integration_tests/allocated_03.f90 [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    implicit none
+    real(4), allocatable :: x(:)
+
+    allocate(x(10))
+    
+    if (allocated(x)) then
+        print *, "x is allocated"
+    else 
+        print *, "x is not allocated"
+    end if
```
Comment (certik) [inline integration_tests/allocated_03.f90:12]: add a check

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 3]
```diff
@@ -7358,6 +7358,14 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         tmp = builder->CreateOr(arg1, arg2);
     }
 
+    void handle_allocated(const ASR::FunctionCall_t& x) {
+        LCOMPILERS_ASSERT(x.n_args == 1);
+        ASR::call_arg_t arg = x.m_args[0];
+        ASR::Variable_t* arg_var = ASRUtils::EXPR2VAR(arg.m_value);
+        fetch_var(arg_var);
+        tmp = arr_descr->get_is_allocated_flag(tmp);
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:7366]: @czgdp1807 do you know if allocatable array always has a descriptor in our LLVM backend?

Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:7366]: Yes they always do. See,

https://github.com/lfortran/lfortran/blob/6eeb126cde6dcc4a1e73371e0fa3c7a010324b7f/src/libasr/asr_utils.h#L3139-L3144

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 4]
```diff
@@ -7543,6 +7551,10 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                 handle_bitwise_or(x);
                 return ;
             }
+            if( startswith(symbol_name, "allocated") ){
+                handle_allocated(x);
+                return ;
+            }
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:7557]: This isn't giving me very good vibes. @certik Can we use `IntrinsicFunction` to represent `allocated` (no need to define `instantiate_Allocated` though, just implement it in respective backends.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 3 above)
Comment (certik) [reply src/libasr/codegen/asr_to_llvm.cpp:7366]: Perfect.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 4 above)
Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:7557]: Makes sense.

> I don't like any of the `startswith(symbol_name, "_bitwise_or")`, that's a bad design.

True.

**Review by certik — APPROVED** [review]
I think that this is fine, thanks!

**Pull request merged**